### PR TITLE
GH-15098: [C++] fix util::EqualityComparable to compile on clang 15

### LINKED ISSUE #15098: [C++] FieldRef with clang-cl from VS 17.4.3 needs operator== with /std:c++20
### Describe the bug, including details regarding any error messages, version, and platform.

Hi

Compiling
```
#include <arrow/api.h>
```
with `clang-cl test.cpp -I/path/to/arrow /std:c++20`

fails:
```
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\xutility(4588,14): error:
      no matching function for call to object of type 'std::equal_to<>'
        if (!_Pred(*_UFirst1, *_UFirst2)) {
             ^~~~~
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\xutility(4605,17): note:
      in instantiation of function template specialization 'std::equal<const arrow::FieldRef
      *, const arrow::FieldRef *, std::equal_to<>>' requested here
    return _STD equal(_First1, _Last1, _First2, equal_to<>{});
                ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\vector(2293,21): note:
      in instantiation of function template specialization 'std::equal<const arrow::FieldRef
      *, const arrow::FieldRef *>' requested here
        return _STD equal(_Left._Unchecked_begin(), _Left._Unchecked_end(), _Right._U...
                    ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\xstddef(213,43): note:
      in instantiation of function template specialization 'std::operator==<arrow::FieldRef,
      std::allocator<arrow::FieldRef>>' requested here
        return static_cast<_Ty1&&>(_Left) == static_cast<_Ty2&&>(_Right);
                                          ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\variant(1326,20): note:
      in instantiation of function template specialization 'std::equal_to<>::operator()<const

      std::vector<arrow::FieldRef> &, const std::vector<arrow::FieldRef> &>' requested here
            return _Op{}(_Variant_raw_get<_Idx>(_Left), _Right._Val);
                   ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\variant(669,23): note:
      in instantiation of function template specialization
      'std::_Variant_relop_visitor2<std::equal_to<>, bool, arrow::FieldPath,
      std::basic_string<char>,
      std::vector<arrow::FieldRef>>::operator()<std::vector<arrow::FieldRef>, 2ULL>'
      requested here
        _STL_STAMP(4, _STL_VISIT_STAMP);
                      ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\variant(714,44): note:
      in instantiation of function template specialization
      'std::_Variant_raw_visit1<1>::_Visit<std::_Variant_relop_visitor2<std::equal_to<>,
      bool, arrow::FieldPath, std::basic_string<char>, std::vector<arrow::FieldRef>>, const
      std::_Variant_storage_<false, arrow::FieldPath, std::basic_string<char>,
      std::vector<arrow::FieldRef>> &>' requested here
    return _Variant_raw_visit1<_Strategy>::_Visit(_Idx, static_cast<_Fn&&>(_Func), st...
                                           ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\variant(1340,12): note:
      in instantiation of function template specialization 'std::_Variant_raw_visit<const
      std::_Variant_storage_<false, arrow::FieldPath, std::basic_string<char>,
      std::vector<arrow::FieldRef>> &, std::_Variant_relop_visitor2<std::equal_to<>, bool,
      arrow::FieldPath, std::basic_string<char>, std::vector<arrow::FieldRef>>>' requested
      here
        && _Variant_raw_visit(_Right_index, _Right._Storage(), _Visitor{_Left._Storage()});
           ^
cpp/src\arrow/type.h(1729,59): note: in instantiation of function template specialization
      'std::operator==<arrow::FieldPath, std::basic_string<char>,
      std::vector<arrow::FieldRef>>' requested here
  bool Equals(const FieldRef& other) const { return impl_ == other.impl_; }
                                                          ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.34.31933\include\xstddef(210,31): note:
      candidate template ignored: substitution failure [with _Ty1 = const arrow::FieldRef &,
      _Ty2 = const arrow::FieldRef &]: ISO C++20 considers use of overloaded operator '=='
      (with operand types 'const arrow::FieldRef' and 'const arrow::FieldRef') to be
      ambiguous despite there being a unique best viable function
    _NODISCARD constexpr auto operator()(_Ty1&& _Left, _Ty2&& _Right) const
```
I don't recall seeing this previously, so maybe a change with the 17.4(.3?) release. variant impl, maybe.

A simple `bool operator==(const FieldRef& other) const { return Equals(other); }` added to `FieldRef` fixes the compilation. I can made a PR if you like.

Thanks

### Component(s)

C++

### PR DESCRIPTION
**Pull request by westonpace**
If you have an equality operator then C++20 then clang 15 expects it to be symmetric.  The util::EqualityComparable operator== was defined as:

`bool operator==(util::EqualityComparable<T>, T)`

and there is no equivalent:

`bool operator==(T, util::EqualityComparable<T>)`

This was later determined to be unintentional strictness and the [standard was clarified](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2468r2.html).  This error was relaxed in clang 16.  Equality only needs to be symmetric if you want the compiler to provide `!=` for you (which we don't rely on).  However, it seems like a worthwhile fix, and it doesn't appear that Clang will be backporting the fix to clang 15 (that is my interpretation of the fact that clang lists the defect as clang-16 in this conformance page: https://clang.llvm.org/cxx_status.html).

Furthermore, the old definition allowed for `Scalar::operator==(std::shared_ptr<Scalar> other)` (i.e. comparing with a shared_ptr) and it isn't clear to me that should be allowed (it's generally inconsistent with the rest of the code base).  This change seems to have undone that and I made no change to restore it (and instead fixed the references to that odd equality operation).

BREAKING CHANGE: The functions `bool Scalar::Equals(const std::shared_ptr<Scalar>&) const` and `bool Scalar::Equals(const std::shared_ptr<Scalar>&, const  EqualOptions&) const` have been removed.  Instead `bool Scalar::Equals(const Scalar&) const` and `bool Scalar::Equals(const Scalar&, const EqualOptions&) const` should be used.

BREAKING CHANGE: `FileInfo::ToString` now includes the size and mtime.

* Closes: #15098

### CODE REVIEW

**Comment by westonpace** [comment]
@lukester1975 made some good efforts to get a reproduction going on Windows.  However, there were some problems.  I've noticed that I can also reproduce this on clang / ubuntu and so I will try and get a CI job setup for C++20 with clang.

**Comment by westonpace** [comment]
Will rerun tests once the github issue is resolved.

**Comment by mapleFU** [comment]
Since I previous tries it but failed to compile under msvc...
@frederick-vs-ja Mind take a look?

**Comment by lukester1975** [comment]
Just to add a few more bits of info:

* I hadn't noticed this with clang on Linux as I was using arrow v9 there
* arrow v10 and any clang 15 version seems to show the issue (but only a warning with clang; hard error with clang-cl)
* Clang 16 has https://github.com/llvm/llvm-project/commit/38b9d313e6945804fffc654f849cfa05ba2c713d
* I have a clang 16 build on Linux and indeed there's no warning raised. I'm not going to try building clang-cl 16, though 😀

Still, no doubt you want a warning free experience for clang 15 (or error free for clang-cl) users!

Thanks

**Comment by mapleFU** [comment]
@lukester1975 According to https://github.com/apache/arrow/pull/13874#issuecomment-1381490629 . This was fixed in clang-16 and gcc-13

**Comment by lukester1975** [comment]
Ah sorry, hadn't seen that other issue.

But yeah, that's what I was saying (well, fixed in clang 16 - gcc 12.2.1 from Fedora 37 is fine for me, so maybe the change was back ported?).

The "real" issue is it seems to be a compilation failure rather than warning with clang-cl 15, irrespective of `/WX`...

Thanks!

**Comment by westonpace** [comment]
I've rebased so the tests will run again and I'll see how close we are.  If I can get this to pass without much more effort it still seems like it will be a good thing.  For example, I think allowing `Scalar == std::shared_ptr<Scalar>` is a bit permissive and unusual and that sort of thing tends to lead to surprises and bugs.  Clang-15 support would be a nice bonus.

**Comment by westonpace** [comment]
@pitrou @bkietz curious for a second opinion on this change.  Also, do we want to keep the `Scalar/shared_ptr<Scalar>` comparison function but mark it as deprecated?

**Comment by mapleFU** [comment]
Any updates? Seems it blocked here...?

**Comment by kou** [comment]
@lidavidm Could you take a look at this?

**Comment by lidavidm** [comment]
Will do, though from first glance I would lean more towards keep-but-deprecate for now

**Review by lidavidm — APPROVED** [review]
Ok, this looks reasonable. I had thought this would affect all types but I suppose only Scalar had the odd equality operator.

**Comment by westonpace** [comment]
I've rebased and will merge once the tests pass.

**Pull request merged**
